### ISSUE TITLE
The order of `access_log` handler settings may not be the best.

### ISSUE THREAD

**Issue by wangfakang** [envoyproxy/envoy#14767]
*Title*: *The order of `access_log` handler settings may not be the best*

*Description*:
Now `access_log` is always executed before each filter log phase of HTTP, which will cause some operations in the log phase of http filter to not take effect in the access_log handler.
Forexample, if the metadata  `xxx :yy` is set in the log phase of HTTP filter, then `%DYNAMIC_METADATA(xxx :yy)%` in access_log format does not work.

[optional *Relevant Links*:]

analysis
====

`access_log` is registered at the [ActiveStream](https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L593).

```
ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connection_manager,
                                                  uint32_t buffer_limit) {
  ……
  for (const AccessLog::InstanceSharedPtr& access_log : connection_manager_.config_.accessLogs()) {
    filter_manager_.addAccessLogHandler(access_log); 
  }
  ……
}

```

 each filter log handler of HTTP is registered at the [decodeHeaders](https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L999).
```
void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapPtr&& headers,
                                                        bool end_stream) {
……
 const bool upgrade_rejected = filter_manager_.createFilterChain() == false;   // will call filter_manager_.addAccessLogHandler
……
}
```

The log handler is registered using [push_back](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L470), So it means FIFO.
```
void FilterManager::addAccessLogHandler(AccessLog::InstanceSharedPtr handler) {
  access_log_handlers_.push_back(handler);
}
```

expect
====
So the `access_log` handler may be better to execute after all log handler of http filter.

**Comment by wangfakang** [comment]
@mattklein123 @htuch Plz review it, I can pull PR if that's ok.

**Comment by htuch** [comment]
Checking my understanding. You are saying that the per-filter access log handlers run _after_ the access log sink handlers registered with HCM?

It seems reasonable for them to run before, but it seems kind of hacky to be mutating stream state at this point when the stream is undergoing destruction.

**Comment by wangfakang** [comment]
> Checking my understanding. You are saying that the per-filter access log handlers run _after_ the access log sink handlers registered with HCM?

Yeah.  per-filter-log and access-log handler should be runs in the same order as  the `p2` picture.

![image](https://user-images.githubusercontent.com/8215757/105623265-6bdeeb80-5e53-11eb-9920-c212a8c90f44.png)

> It seems reasonable for them to run before, but it seems kind of hacky to be mutating stream state at this point when the stream is undergoing destruction.

Usually we might need to do something in the filter log phase and then log it via the `access_log` handler, but this is invalid now. Forexample we create the metadata `xxx :yy` is set in the log phase of HTTP filter, then the %DYNAMIC_METADATA(xxx :yy)% in access_log format does not work.

So I propose that the runtime of the `access_log` handler should be placed at the end of the per-filter log,
Just like the picture `p2` above.

**Comment by wangfakang** [comment]
ping @htuch

**Comment by htuch** [comment]
This seems reasonable, although I still don't fully grok why you do dynamic metadata creation during destruction.

**Comment by wangfakang** [comment]
> This seems reasonable, although I still don't fully grok why you do dynamic metadata creation during destruction.

@htuch  Forexample, we need to unify the processing time of some http filter, and then record the statistical results to the access log through dynamic metadata.

**Comment by TAOXUY** [comment]
+1 on this request

**Comment by spacewander** [comment]
@htuch @mattklein123 
What about adding a new method `prependAccessLogHandler`? Adding this method won't break any code. If a filter wants to record the statistical results to the access log through dynamic metadata, it can choose the new method. The `addAccessLogHandler` can still be used by the HCM to insert multiple access log handlers in the current order.

This also solves #30859 easily: Just change one line `addAccessLogHandler` to `prependAccessLogHandler`.

**Comment by htuch** [comment]
Sounds reasonable to me, @mattklein123?

**Comment by Elliot-xq** [comment]
Hey @spacewander, could you elaborate more on the `prependAccessLogHandler`? is `prependAccessLogHandler` still adding loggers to the [access_log_handlers_](https://source.corp.google.com/piper///depot/google3/third_party/envoy/src/source/common/http/filter_manager.h;bpv=1;bpt=1;l=1034?q=access_log_handlers_&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental)? Why could it ensure that access_logger logs after the filter log?

to provide some context: https://github.com/envoyproxy/envoy/issues/30859 seems to be still needed for us. We would love to make the `prependAccessLogHandler` changes if it could solve our problem.

**Comment by mattklein123** [comment]
Sure sounds good to me.

**Comment by spacewander** [comment]
> Hey @spacewander, could you elaborate more on the `prependAccessLogHandler`? is `prependAccessLogHandler` still adding loggers to the [access_log_handlers_](https://source.corp.google.com/piper///depot/google3/third_party/envoy/src/source/common/http/filter_manager.h;bpv=1;bpt=1;l=1034?q=access_log_handlers_&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental)? Why could it ensure that access_logger logs after the filter log?
> 
> to provide some context: Seems we still require fix for #30859 . We would love to make the `prependAccessLogHandler` changes if it could solve our problem.

@Elliot-xq 

Yes, `prependAccessLogHandler` still adds loggers to the `access_log_handlers_`, just in a different direction.

The current status is:
1. Envoy adds the access log handler by appending it to the end
2. the access_loggers are added before the filter loggers

So if we provide a new method to add an access log handler by prepending it, we can use this method to add the filter log before the access_loggers, without changing the execution order (just change the method we call in the filter).

I am also going to change the [golang filter's access log](https://github.com/envoyproxy/envoy/blob/b5bbfba79379eaf4824c3a96c4d9d19bd498a31a/contrib/golang/filters/http/source/config.cc#L47), which is a part of my recent job to improve its access log mechanism. Hope this would be helpful to you.

**Comment by Elliot-xq** [comment]
Thanks for the explanation @spacewander !!
